### ISSUE TITLE
libmp3lame cannot be linked dynamically in new CMake build system

### ISSUE THREAD

**Issue by mikhailnov** [audacity/audacity#554]
In the new cmake build system it is impossible to link libmp3lame dynamically. It was possible in autotools, and the code of Audacity is designed to allow to work without dlopening it.

cmake-proxies/lame assumes it to be linked statically and src/CMakeLists.txt sets DISABLE_DYNAMIC_LOADING_LAME to YES only in Windows builds. src/audacity_config.in default to DISABLE_DYNAMIC_LOADING_LAME being YES.

In ffmpeg, audacity_use_ffmpeg:STRING=linked makes it be linked dynamically, but why do you use "linked" for ffmpeg, why not just link it dynamicaly if a system version is used? And why is lame not using the same config options?

**Comment by mikhailnov** [comment]
`-Daudacity_use_ffmpeg:STRING=linked` does not set DISABLE_DYNAMIC_LOADING_FFMPEG, with that build option ffmpeg is still dlopened:
`openat(AT_FDCWD, "/usr/lib64/libavcodec.so.58", O_RDONLY|O_CLOEXEC) = 3`

**Comment by mikhailnov** [comment]
Seems that -DDISABLE_DYNAMIC_LOADING_LAME is set, but everything continues to work because some of dependencies of Audacity depend from libmp3lame, what makes it already loaded.
libmp3lame must be explicitly linked to Audacity.

**Comment by rrrapha** [comment]
> In the new cmake build system it is impossible to link libmp3lame dynamically.

Yes, currently no system version of lame is detected.
`pkg_check_modules` doesn't work here because mp3lame doesn't provide a pkg-config file.

https://github.com/audacity/audacity/blob/ce07da936cf956c2d2226c356d287e32f4323aad/cmake-proxies/CMakeLists.txt#L86

**Comment by plater** [comment]
This is a blocker for updating to 2.4.2 in openSUSE and most probably other distributions too. Looking at @rrrapha 's statement I'm led to believe that if I create a pkgconfig/mp3lame.pc in our llibmp3lame package that cmake will pick it up?
It's a pity that autotools was removed prematurely.

**Comment by plater** [comment]
The actual library is called libmp3lame but audacity's cmake looks for lame.pc not libmp3lame.pc if this is changed I'll create a pkgconfig file in the lame package, meanwhile I'll make a workaround for the online build until either system lame is found like the old autotools did (preferable) or cmake looks for libmp3lame.pc

**Comment by lllucius** [comment]
> 
> 
> This is a blocker for updating to 2.4.2 in openSUSE and most probably other distributions too. Looking at @rrrapha 's statement I'm led to believe that if I create a pkgconfig/mp3lame.pc in our llibmp3lame package that cmake will pick it up?
> It's a pity that autotools was removed prematurely.

If this is blocker, then you're going to LOVE our next release.  We will be requiring use of our modified version of wxWidgets (https://github.com/audacity/wxWidgets)  This would have been in 2.4.2, but I didn't make the deadline.

**Comment by mikhailnov** [comment]
We will love it! What are the reasons to require a modified version of wxWidgets?

P.S. I hacked Audacity to use system libmp3lame with a patch bellow and by adding `-lmp3lame` to linker flags manually.

```diff
From f0c7f63e59c2db9183908be556836e77e4e0755b Mon Sep 17 00:00:00 2001
From: Mikhail Novosyolov <mikhailnov@nixtux.ru>
Date: Sun, 31 May 2020 02:36:51 +0300
Subject: [PATCH] HACK: off bundled libmp3lame

https://github.com/audacity/audacity/issues/554
Manually setting
-DDISABLE_DYNAMIC_LOADING_LAME=1 -lmp3lame
---
 cmake-proxies/lame/CMakeLists.txt | 26 +-------------------------
 lib-src/lame/empty.c              |  3 +++
 2 files changed, 4 insertions(+), 25 deletions(-)
 create mode 100644 lib-src/lame/empty.c
diff --git a/cmake-proxies/lame/CMakeLists.txt b/cmake-proxies/lame/CMakeLists.txt
index ac405a6d5..f521fdddb 100644
--- a/cmake-proxies/lame/CMakeLists.txt
+++ b/cmake-proxies/lame/CMakeLists.txt
@@ -6,26 +6,7 @@ def_vars()
 list( APPEND SOURCES
    PRIVATE
       # libmp3lame
-      ${TARGET_ROOT}/libmp3lame/bitstream.c
-      ${TARGET_ROOT}/libmp3lame/encoder.c
-      ${TARGET_ROOT}/libmp3lame/fft.c
-      ${TARGET_ROOT}/libmp3lame/gain_analysis.c
-      ${TARGET_ROOT}/libmp3lame/id3tag.c
-      ${TARGET_ROOT}/libmp3lame/lame.c
-      ${TARGET_ROOT}/libmp3lame/newmdct.c
-      ${TARGET_ROOT}/libmp3lame/presets.c
-      ${TARGET_ROOT}/libmp3lame/psymodel.c
-      ${TARGET_ROOT}/libmp3lame/quantize.c
-      ${TARGET_ROOT}/libmp3lame/quantize_pvt.c
-      ${TARGET_ROOT}/libmp3lame/reservoir.c
-      ${TARGET_ROOT}/libmp3lame/set_get.c
-      ${TARGET_ROOT}/libmp3lame/tables.c
-      ${TARGET_ROOT}/libmp3lame/takehiro.c
-      ${TARGET_ROOT}/libmp3lame/util.c
-      ${TARGET_ROOT}/libmp3lame/vbrquantize.c
-      ${TARGET_ROOT}/libmp3lame/VbrTag.c
-      ${TARGET_ROOT}/libmp3lame/version.c
-      ${TARGET_ROOT}/libmp3lame/vector/xmm_quantize_sub.c
+      ${TARGET_ROOT}/empty.c
 )
 
 list( APPEND INCLUDES
@@ -82,8 +63,3 @@ configure_file( config.h.in private/config.h )
 
 #organize_source( "${TARGET_ROOT}" "" "${SOURCES}" )
 target_sources( ${TARGET} PRIVATE ${SOURCES} )
-target_compile_definitions( ${TARGET} PRIVATE ${DEFINES} )
-target_include_directories( ${TARGET} PRIVATE ${INCLUDES} )
-target_compile_options( ${TARGET} PRIVATE ${OPTIONS} )
-target_link_libraries( ${TARGET} PRIVATE ${LIBRARIES} )
-
diff --git a/lib-src/lame/empty.c b/lib-src/lame/empty.c
new file mode 100644
index 000000000..292aa3b34
--- /dev/null
+++ b/lib-src/lame/empty.c
@@ -0,0 +1,3 @@
+int main(){
+	//
+} 
```

**Comment by lllucius** [comment]
Commits for our next version have been opened back up, so I'll look into fixing this for you guys.

The reason for forcing the custom version is just that, it's custom.  We have backported several important fixes from wxWidgets master that Audacity needs.  And we'll be adding Audacity specific accessibility to our wxWidgets version.  (Though I hope to make it generic enough to submit upstream.)

**Comment by plater** [comment]
I've added a pkgconfig file to the lame package and added a generated pc file to audacity's spec file. If there's a patch available for the wxWidgets modification Audacity already has it's own widgets package in openSUSE. Which commit can I get the widgets patch from?

**Comment by plater** [comment]
BTW a little off topic, audacity no longer builds for 32 bit.

**Comment by lllucius** [comment]
> 
> 
> I've added a pkgconfig file to the lame package and added a generated pc file to audacity's spec file. If there's a patch available for the wxWidgets modification Audacity already has it's own widgets package in openSUSE. Which commit can I get the widgets patch from?

We keep our version at https://github.com/audacity/wxWidgets and the default branch is set to audacity-fixes-3.1.3.  We will always have our current version set as the default branch in Github, so just cloning will get you the most current.

**Comment by lllucius** [comment]
> 
> 
> BTW a little off topic, audacity no longer builds for 32 bit.

Interesting, it does on Windows.  What errors are you getting?  That's something we need to fix.

**Comment by mikhailnov** [comment]
@lllucius So in general audacity's version of wxWidgets does not break compatibility with upstream wxWidgets and applications using it, does it?

**Comment by plater** [comment]
I made a mistake with the generated pc file and hardcoded lib64, this is most probably the cause

**Comment by plater** [comment]
The build failure for i586 linux has been reported in https://github.com/audacity/audacity/issues/594

**Comment by JamesCrook** [comment]
Closing per:
> I made a mistake with the generated pc file and hardcoded lib64, this is most probably the cause

**Comment by plater** [comment]
> > BTW a little off topic, audacity no longer builds for 32 bit.
> 
> Interesting, it does on Windows. What errors are you getting? That's something we need to fix.

See https://github.com/audacity/audacity/issues/594#issuecomment-840788325

**Comment by mikhailnov** [comment]
I think this issue should be reopened

**Comment by JamesCrook** [comment]
> I think this issue should be reopened

Done.

**Comment by crsib** [comment]
Fixed in #1008
